### PR TITLE
[Core] Register multiple key events/presses per USB report

### PR DESCRIPTION
**Pull request by KarlK90**
## Description

This is a continuation of #10955 by @hongaaronc (great work!) and addresses open issues with the implementation. Namely tapped key codes and shifted key codes.  Please see the original PR for discussion and issues related to it.
 
#### Changes to adapt for tapped and subsequent shifted key events

Tapped key events by default work only on **key release** by sending a key press event and a subsequent key release event to the host. Now with deferring the send event after the complete matrix scan these events are merged and erase each other becoming effectively a nop. The idea is to buffer the unregister event of the key release for all tapped keys (normal keys are not buffered) and send these in a separate keyboard report just after the key press events have been send.

Unregister events/key codes are stored in the `unregister_keycodes` struct by `unregister_code_buffered()` which is called for all tapped key codes and send to the host after a complete scan is completed by `send_keyboard_report_buffered_unregister_keys()`.

Subsequent shifted key codes like pressing `KC_EQL` and `KC_PLUS` (`KC_LSHIFT` + `KC_EQL`) while holding the former need an additional key release event of `KC_EQL` before sending the combination again. This release event isn't deferred anymore but send as a report immediately.

All failing unit-tests have been fixed or rather adapted for this feature. As it turns out most of the failing key press cases expect one report per key event so they are false negatives. ATM the adapted unit-tests are behind a `#if defined` guard, I'm not familiar with the CI infrastructure and this needs addressing, currently two binaries would have to be compiled one with the default processing enabled and one with `REGISTER_MULTIPLE_KEYEVENTS_ENABLE`.

`DEFER_KEYBOARD_REPORT_ENABLE` was renamed to `REGISTER_MULTIPLE_KEYEVENTS_ENABLE` because it rather states a implementation detail but doesn't communicate purpose very well in my opinion.

## Types of Changes

- [x] Core
- [x] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Keys at the same time #4904
* Apparently the community of the keyboard rhythm games [OSU and Etterna warns about chord splitting when using QMK](https://www.youtube.com/watch?v=heZVmr9fyng&t=282s) powered keyboards there is also a [spreadsheet](https://docs.google.com/spreadsheets/d/1hNMXn_nkpXZIpDHZBcp95E7OweL4594_WDLCY_vTvm0/edit#gid=0) about the phenomena.
* Multiple mentions of bugs / quirks [0](https://github.com/qmk/qmk_firmware/blob/5e98eaaaff8fde1ce25b9bad6c00a982718cb467/tests/basic/test_keypress.cpp#L45-L46) [1](https://github.com/qmk/qmk_firmware/blob/5e98eaaaff8fde1ce25b9bad6c00a982718cb467/tests/basic/test_keypress.cpp#L72-L73) [2](https://github.com/qmk/qmk_firmware/blob/5e98eaaaff8fde1ce25b9bad6c00a982718cb467/tests/basic/test_keypress.cpp#L170-L173) [3](https://github.com/qmk/qmk_firmware/blob/5e98eaaaff8fde1ce25b9bad6c00a982718cb467/tests/basic/test_keypress.cpp#L219-L222) mentioned in the keypress unit-tests are solved by this.

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by daskygit** [comment]
This may be intentional, but this breaks `TAP_CODE_DELAY` when it's enabled with `REGISTER_MULTIPLE_KEYEVENTS = yes` in rules.mk

I have the polling rate set to 1000hz so taps from `LT(_LOWER, KC_TAB)` have a very short period between TAB registering and unregistering and some applications don't play nicely with such a short keypress.

Setting `#define TAP_CODE_DELAY 5` increases the delay before the unregister event occurs but with this feature enabled it appears to be ignored. 

If I disable the feature with `REGISTER_MULTIPLE_KEYEVENTS = no` in rules.mk `TAP_CODE_DELAY` works as expected.

I tested the following keycodes:
`LT(_LOWER, KC_TAB)`
`MT(MOD_LCTL, KC_NUBS)`

**Comment by KarlK90** [comment]
Thanks for the feedback!

The breakage is not intentional, I have added tap delays to the unregister process and those keycodes should be working again!

File: tmk_core/common/action_util.h [hunk 1]
```diff
@@ -18,22 +18,45 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
 #include <stdint.h>
+#include <stddef.h>
 #include "report.h"
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-extern report_keyboard_t *keyboard_report;
+#if !defined(QMK_KEYS_PER_SCAN) && defined(REGISTER_MULTIPLE_KEYEVENTS_ENABLED)
```
Comment (daskygit) [inline tmk_core/common/action_util.h:28]:
```suggestion
#if !defined(QMK_KEYS_PER_SCAN) && defined(REGISTER_MULTIPLE_KEYEVENTS_ENABLE)
```

File: tmk_core/common/action_util.h [hunk 2]
```diff
@@ -18,22 +18,45 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+#    warning REGISTER_MULTIPLE_KEYEVENTS_ENABLED needs QMK_KEYS_PER_SCAN defined to have any effect, setting default of 4.
```
Comment (daskygit) [inline tmk_core/common/action_util.h:29]:
```suggestion
```

Warning are treated as errors and stops the compilation (at least by default). If it can just define a decent default does it need to show a warning?

**Comment by daskygit** [comment]
I had a nice long message typed out about a issue I was having but it made me think a little, it was partly user error. I didn't have `QMK_KEYS_PER_SCAN` defined but noticed it should be defaulting to 4 and I wasn't seeing a warning message.

File: tmk_core/common/action_util.h (hunk 2 above)
Comment (KarlK90) [reply tmk_core/common/action_util.h]: Converted to `#pragma message`

File: tmk_core/common/action_util.h (hunk 1 above)
Comment (KarlK90) [reply tmk_core/common/action_util.h]: Fixed.

**Comment by KarlK90** [comment]
Fixed both the typo and the warning. Thank you for the feedback.

**Comment by daskygit** [comment]
The stale bot brought me here, so just thought I'd add I've been using this for sometime now and not noticed any issues. Thanks again.

**Review by XScorpion2 — APPROVED** [review]

File: tmk_core/common/action.c [hunk 3]
```diff
@@ -889,9 +938,13 @@ void unregister_code(uint8_t code) {
  */
 void tap_code_delay(uint8_t code, uint16_t delay) {
     register_code(code);
+#if defined(__AVR__)
```
Comment (drashna) [inline tmk_core/common/action.c:941]: Is there a specific reason for this change?

Comment (KarlK90) [reply tmk_core/common/action.c:941]: Waiting on the Chibios platform is done by de-scheduling the current thread and re-scheduling after the waiting time. Under the hood this also means setting a wakeup timer and reacting to this timer interrupt. In bulk this is rather costly, so I merged this into one continuous waiting block.

**Comment by drashna** [comment]
Looks like there are some merge conflicts here that need to be resolved.

File: tmk_core/common/action.c (hunk 3 above)
Comment (zvecr) [reply tmk_core/common/action.c:941]: Im not a fan of this change, as it bleeds more platform knowledge into code code (and from what i can tell, isnt strictly required for the overall aim of this PR).

Personally I would like to see it PRed separately, maybe hiding the implementation difference within the wait_ms AVR code.

Comment (KarlK90) [reply tmk_core/common/action.c]: Reverted, I'm going to open a separate PR once this is merged.

**Comment by KarlK90** [comment]
Rebased my commits onto develop and squashed my commits into a single one. Should be good to merge now.

**Comment by drashna** [comment]
This definitely causes issues for me.  

I'm not sure about other keys, but oneshot mods... if I hit both the OSM key and a regular key at the same time, neither are registered.  Disabling this option immediately fixes the issue. 

I'm not sure if this effects LT/MT keys too, but I suspect that it may, depending on the tap-hold config options set.

**Comment by KarlK90** [comment]
Oh well, I guess this is the point where more unit-tests have to be added to spot regressions like this.

**Comment by KarlK90** [comment]
@drashna I have added some more one shot modifier and one shot layer unit-tests that pass on develop and only without the changes of this pr. BUT the default behavior of QMK seems just wrong (if this is not due to my unit-test logic). For instance QMK sends [8 usb reports](https://github.com/qmk/qmk_firmware/pull/12686/commits/62e013155090a03c529e147991252b1f26c28686#diff-dea0c43ef0acdef01736c5cc7f96aa64838ba4c57742148c66eedcdf659795b7R201-R228) for a simple OSL + regular key combo.

**Comment by drashna** [comment]
> I'm wondering if this behavior comes down to the unit-tests and reality is saner or if this feature just unearthed some buggy behavior that was unnoticed before.

Well, I use one shot mods extensively, and it's not behavior I've had before, even using the qmk scan define.   However, it is possible that the code change here revealed an edge case.

**Comment by KarlK90** [comment]
I'm pulling the addition of unit-tests out of this PR into a separate one #13789 as those tests should be written against the current gold standard aka. the develop branch.

**Comment by drashna** [comment]
Testing this out again, and it works much better. I do use a combination of mod tap, layer tap, and one shot mods, so a good sampling of behavior.  

And this looks good.

**Comment by KarlK90** [comment]
But I haven't changed anything in the core code :thinking:, was it an old version of this pr?

**Comment by drashna** [comment]
Hmm, it may be other code changes that are causing issues or otherwise amplifying issues.  
I did clean out my branch, so that may be why, it's behaving better. 

But now that you've mentioned it, I still do see the issue, and with mod tap, but nowhere near as frequently, and it's very difficult to trigger, right now. But it does still happen.  Will

File: tmk_core/common/action.c [hunk 4]
```diff
@@ -784,49 +792,86 @@ void register_code(uint8_t code) {
...
+#else
+
+    if (delay > 0) {
+#    if defined(__AVR__)
+        for (uint16_t i = delay; i > 0; i--) {
+            wait_ms(1);
+        }
+#    else
+        wait_ms(delay);
+#    endif
```
Comment (zvecr) [inline tmk_core/common/action.c:863]:
```suggestion
        wait_ms(delay);
```

File: tmk_core/common/action_util.c [hunk 5]
```diff
@@ -252,11 +252,62 @@ void send_keyboard_report(void) {
...
+            unregister_keycodes.len--;
+            unregister_code_deferred(unregister_keycodes.buffer[unregister_keycodes.len]);
+        }
+
+        if (unregister_keycodes.tap_delay != 0) {
+#if defined(__AVR__)
+            for (uint16_t i = unregister_keycodes.tap_delay; i > 0; i--) {
+                wait_ms(1);
+            }
+#else
```
Comment (zvecr) [inline tmk_core/common/action_util.c:301]:
```suggestion
```

File: tmk_core/common/action_util.c [hunk 6]
```diff
@@ -252,11 +252,62 @@ void send_keyboard_report(void) {
...
+            wait_ms(unregister_keycodes.tap_delay);
 #endif
```
Comment (zvecr) [inline tmk_core/common/action_util.c:303]:
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]
With #14025, the AVR specific logic is no longer required.

**Comment by KarlK90** [comment]
@zvecr Thank you!

**Comment by daskygit** [comment]
Just going to give this a little bump, been using this for quite some time without no noticeable problems. Merging it early this cycle would give it plenty of extra testing time. :wink:

File: quantum/action_util.h [hunk 7]
```diff
@@ -18,22 +18,45 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
 #endif
 
-extern report_keyboard_t *keyboard_report;
+#if !defined(QMK_KEYS_PER_SCAN) && defined(REGISTER_MULTIPLE_KEYEVENTS_ENABLE)
+#    pragma message "REGISTER_MULTIPLE_KEYEVENTS_ENABLE needs QMK_KEYS_PER_SCAN defined to have any effect, setting default of 4."
+#    define QMK_KEYS_PER_SCAN 4
+#endif
+
+#if !defined(QMK_KEYS_PER_SCAN)
+#    define UNREGISTER_KEYCODES_BUFFER_SIZE 1
```
Comment (drashna) [inline quantum/action_util.h:34]: not documented?

Comment (drashna) [reply quantum/action_util.h:34]: Not configurable?

**Comment by drashna** [comment]
> Just going to give this a little bump, been using this for quite some time with no noticeable problems. Merging it early this cycle would give it plenty of extra testing time. 😉

'm consistently having problems with this, actually. 

Se previos statements for... n/m.... yeah.  You can see exactly what 'm talking about.

**Comment by KarlK90** [comment]
I would very much like to merge https://github.com/qmk/qmk_firmware/pull/13789 first, afterwards I'll rebase and continue with these issues with proper unit-tests.

File: quantum/action_util.h (hunk 7 above)
Comment (KarlK90) [reply quantum/action_util.h:34]: Actually I'll would like to drop `QMK_KEYS_PER_SCAN` entirely and give the main keyboard loop a facelift :-), will add documentation nevertheless.

**Comment by ecerulm** [comment]
How do you run the tests, I tried with `make test:basic DEBUG=1 CC=gcc-11 CFLAGS="-DREGISTER_MULTIPLE_KEYEVENTS_ENABLE=1"`  but it will complain : 

```
Linking: .build/test/basic.elf                                                                      [ERRORS]
 | 
 | Undefined symbols for architecture x86_64:
 |   "_timer_elapsed_fast", referenced from:
 |       _debounce in sym_defer_g.o
 |   "_timer_read_fast", referenced from:
 |       _debounce in sym_defer_g.o
 | ld: symbol(s) not found for architecture x86_64
 | collect2: error: ld returned 1 exit status 
``` 

Also I think it would be great if you could a test case that shows how this is different from the current behaviour. Somethinglike this (not tested  and not event sure this is the current behaviour because I can't make it compile with  `QMK_KEYS_PER_SCAN`) : 

```
TEST_F(KeyPress, CorrectKeysAreReportedWhenTwoKeysArePressedX) {
    TestDriver driver;
    press_key(1, 0);
    press_key(0, 3);
   // two keys are pressed how many usb reports will be generated and the contents depends on the features enabled
#if defined(REGISTER_MULTIPLE_KEYEVENTS_ENABLE)
    // single usb report with both keycodes/scancodes
    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_B, KC_C)));
#elif defined (QMK_KEYS_PER_SCAN) 
    // two usb reports, one containing B the first keypress and the second usb report containing both (chord splitting)
    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_B)));
    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_B, KC_C)));
#else
    // two  usb reports each one containing each individual keypress
    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_B)));
    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_C)));
#endif
    keyboard_task();
    release_key(1, 0);
    release_key(0, 3);
    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport()));
    keyboard_task();
}
```

In my opinion this would be good although is more for reference/documenting actual behaviour (more than to test regressions).

**Comment by KarlK90** [comment]
Hi @ecerulm I'm currently working on #15292 which deprecates `QMK_KEYS_PER_SCAN` and makes it the default behavior. Now that we have at least some unit-tests in place I also intend to make the behavior of this PR the default of QMK to reduce the amount of complexity. Some Unit-tests are already in place in this PR but more shall be added of course.

**Comment by ecerulm** [comment]
Looking forward to see this as the standard and only behaviour I don't think there is any need to keep the old behaviour around. 

I think it should be mentioned in the documentation in this PR that QMK does not suffer from "chord splitting" anymore and that is suitable for gaming and for rythm games in particular where pressing multiple keys at once with accurate timing is essential (especially at lower polling rates). So that it's easy to find this information for people that google it.

**Comment by KarlK90** [comment]
Yeah, If you look in the `Issues Fixed or Closed by This PR` section you can see that the original motivation for this PR was people complained that QMK is not suitable for rhythm games. :slightly_smiling_face: I agree, once this has been merged we should promote it more and get the word out there.

**Comment by XScorpion2** [comment]
bump?

**Comment by KarlK90** [comment]
> bump?

Still WIP, I intend to continue with it after #15292

**Comment by openglfreak** [comment]
It looks like this code does not work when the matrix scan rate is significantly higher than the USB polling rate

**Comment by KarlK90** [comment]
> It looks like this code does not work when the matrix scan rate is significantly higher than the USB polling rate

You're right. My (yet to be implemented) approach would be to just cap/couple the matrix scanning frequency to the USB polling frequency. The state of the implementation as it is won't be merged tough. I want to revisit the design choices and make the multiple key events behavior the standard/default one.

**Comment by zerodividesbyyou** [comment]
Bump again?
Would like to know if any progress has been made or if there's an easy way for me to test this with my boards as well

**Comment by longnguyen2004** [comment]
Another bump 4 months later, how is this going now?

**Comment by StrangePeanut** [comment]
Pardon me for being naïve but for anyone subscribed to this PR in anticipation, what is keeping this draft from being merged as is in a separate commit?

**Comment by tzarc** [comment]
> Pardon me for being naïve but for anyone subscribed to this PR in anticipation, what is keeping this draft from being merged as is in a separate commit?

https://github.com/qmk/qmk_firmware/pull/12686#issuecomment-1427112772

**Comment by StrangePeanut** [comment]
@tzarc I get that but it needs to be explicitly enabled and shouldn't affect existing configurations at all. Why couldn't it benefit users who knowingly enable it today?

**Comment by tzarc** [comment]
> @tzarc I get that but it needs to be explicitly enabled and shouldn't affect existing configurations at all. Why couldn't it benefit users who knowingly enable it today?

You’re more than welcome to apply the PR to your own fork if you’re happy with its current state.

The original author is not, and that’s enough reason for it not to go to QMK mainline.

**Comment by amarz45** [comment]
Has there been any work on this as of now?
